### ISSUE TITLE
TypeError if both storage and WilcoxonPruner are given to create_study

### ISSUE THREAD

**Issue by asi1024** [optuna/optuna#5471]
At the following lines, ndarrays of string dtype are accidently created if the keys of trial.intermediate_values are not integers.
https://github.com/optuna/optuna/blob/efe6f3a0d66a5cfd146a609d8a824d1eb0674ccc/optuna/pruners/_wilcoxon.py#L152
https://github.com/optuna/optuna/blob/efe6f3a0d66a5cfd146a609d8a824d1eb0674ccc/optuna/pruners/_wilcoxon.py#L174
I have confirmed that making the following modifications to these lines results in a run without any errors.

```py
steps = np.array(list(trial.intermediate_values.keys()))
step_values = np.array(list(trial.intermediate_values.values()))
```

### Environment

- Optuna version: 3.6.1
- Python version: 3.12.2

### Error messages, stack traces, or logs

```shell
/home/imanishi/er.py:23: ExperimentalWarning: WilcoxonPruner is experimental (supported from v3.6.0). The interface can change in the future.
  pruner=optuna.pruners.WilcoxonPruner(p_threshold=0.1),
[I 2024-05-31 15:21:31,280] A new study created in RDB with name: example-study
5
[I 2024-05-31 15:21:31,348] Trial 0 finished with value: 0.06076577758622594 and parameters: {'a': -0.14410335730379753}. Best is trial 0 with value: 0.06076577758622594.
[W 2024-05-31 15:21:31,376] Trial 1 failed with parameters: {'a': 0.38394546515643824} because of the following error: TypeError("ufunc 'isfinite' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''").
Traceback (most recent call last):
  File "/home/imanishi/.pyenv/versions/3.12.2/lib/python3.12/site-packages/optuna/study/_optimize.py", line 196, in _run_trial
    value_or_values = func(trial)
                      ^^^^^^^^^^^
  File "/home/imanishi/er.py", line 12, in objective
    if trial.should_prune():
       ^^^^^^^^^^^^^^^^^^^^
  File "/home/imanishi/.pyenv/versions/3.12.2/lib/python3.12/site-packages/optuna/trial/_trial.py", line 533, in should_prune
    return self.study.pruner.prune(self.study, trial)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/imanishi/.pyenv/versions/3.12.2/lib/python3.12/site-packages/optuna/pruners/_wilcoxon.py", line 175, in prune
    if np.any(~np.isfinite(best_step_values)):
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: ufunc 'isfinite' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
[W 2024-05-31 15:21:31,377] Trial 1 failed with value None.
Traceback (most recent call last):
  File "/home/imanishi/er.py", line 25, in <module>
    study.optimize(objective, n_trials=300)
  File "/home/imanishi/.pyenv/versions/3.12.2/lib/python3.12/site-packages/optuna/study/study.py", line 451, in optimize
    _optimize(
  File "/home/imanishi/.pyenv/versions/3.12.2/lib/python3.12/site-packages/optuna/study/_optimize.py", line 62, in _optimize
    _optimize_sequential(
  File "/home/imanishi/.pyenv/versions/3.12.2/lib/python3.12/site-packages/optuna/study/_optimize.py", line 159, in _optimize_sequential
    frozen_trial = _run_trial(study, func, catch)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/imanishi/.pyenv/versions/3.12.2/lib/python3.12/site-packages/optuna/study/_optimize.py", line 247, in _run_trial
    raise func_err
  File "/home/imanishi/.pyenv/versions/3.12.2/lib/python3.12/site-packages/optuna/study/_optimize.py", line 196, in _run_trial
    value_or_values = func(trial)
                      ^^^^^^^^^^^
  File "/home/imanishi/er.py", line 12, in objective
    if trial.should_prune():
       ^^^^^^^^^^^^^^^^^^^^
  File "/home/imanishi/.pyenv/versions/3.12.2/lib/python3.12/site-packages/optuna/trial/_trial.py", line 533, in should_prune
    return self.study.pruner.prune(self.study, trial)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/imanishi/.pyenv/versions/3.12.2/lib/python3.12/site-packages/optuna/pruners/_wilcoxon.py", line 175, in prune
    if np.any(~np.isfinite(best_step_values)):
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: ufunc 'isfinite' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
```

### Steps to reproduce

```python
import numpy as np
import optuna


def objective(trial):
    a = trial.suggest_float('a', -1, 1)
    res = []
    for i in np.arange(5):
        loss = a * a + i * 0.02
        res.append(loss)
        trial.report(loss, i)
        if trial.should_prune():
            return np.array(res).mean()
    return np.array(res).mean()


if __name__ == "__main__":
    study = optuna.create_study(
        study_name="example-study",
        storage="sqlite:///db.sqlite3",
        pruner=optuna.pruners.WilcoxonPruner(p_threshold=0.1),
    )
    study.optimize(objective, n_trials=300)
    print(study.best_params)
```

**Comment by contramundum53** [comment]
Reproduced. Thank you!

**Comment by contramundum53** [comment]
It seems that with SQLite, the `trial.intermediate_values` will be something like `{b'\x00\x00\x00\x00\x00\x00\x00\x00': 0.5988937189366361, b'\x01\x00\x00\x00\x00\x00\x00\x00': 0.6188937189366361, b'\x02\x00\x00\x00\x00\x00\x00\x00': 0.6388937189366362, b'\x03\x00\x00\x00\x00\x00\x00\x00': 0.6588937189366362, b'\x04\x00\x00\x00\x00\x00\x00\x00': 0.6788937189366361}`...

**Comment by contramundum53** [comment]
I don't know why CI doesn't fail with such a bug.

**Comment by not522** [comment]
Thank you for your bug report! I found that the error was caused by using `np.int64` instead of `int` as `step`. I should be fixed with `trial.report(loss, int(i))`.
In the current implementation, the reported `value` is converted to a `float`, but the `step` is not converted to an `int`. I think that adding the conversion is good for convenience.
https://github.com/optuna/optuna/blob/efe6f3a0d66a5cfd146a609d8a824d1eb0674ccc/optuna/trial/_trial.py#L481
